libsrc/cx16/tgi/cx640p1.s

```diff
@@ -0,0 +1,677 @@
+
+
+.code
+
+; ------------------------------------------------------------------------
+; INSTALL routine. Is called after the driver is loaded into memory. May
+; initialize anything that has to be done just once. Is probably empty
+; most of the time.
+;
+; Must set an error code: NO
+
+INSTALL:
+; Create the default palette.
+        lda #$00
+        sta defpalette
+        lda #$01
+        sta defpalette+1
+
+        ; Fall through.
+
+; ------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory. May
+; clean up anything done by INSTALL, but is probably empty most of the time.
+;
+; Must set an error code: NO
+
+UNINSTALL:
+        rts
+
+; ------------------------------------------------------------------------
+; INIT: Changes an already installed device from text mode to graphics
+; mode.
+; Note that INIT/DONE may be called multiple times while the driver
+; is loaded, while INSTALL is called only once; so, any code that is needed
+; to initiate variables and so on must go here.  Setting the palette is not
+; needed because that is called by the graphics kernel later.
+; The graphics kernel never will call INIT when a graphics mode already is
+; active, so there is no need to protect against that.
+;
+; Must set an error code: YES
+
+INIT:   stz     ERROR           ; #TGI_ERR_OK
+
+; Save the current text mode.
+
+        sec
+        jsr     SCREEN_MODE
+        sta     text_mode
+
+; Switch into (640 x 480 x 2 bpp) graphics mode.
+
+        lda #%00000000          ; DCSEL = 0, VRAM port 1
+        sta VERA::CTRL
+        lda #%00100001          ; Disable sprites, layer 1 enable, VGA
+        sta VERA::DISP::VIDEO
+        lda #%00000100          ; Bitmap mode enable
+        sta VERA::L1::CONFIG
+        lda #%00000001          ; Tile width 640
+        sta VERA::L1::TILE_BASE
+        rts
+
+; ------------------------------------------------------------------------
+; DONE: Will be called to switch the graphics device back into text mode.
+; The graphics kernel never will call DONE when no graphics mode is active,
+; so there is no need to protect against that.
+;
+; Must set an error code: NO
+
+DONE:
+        jsr     CINT
+        lda     text_mode
+        clc
+        jmp     SCREEN_MODE
+
+; ------------------------------------------------------------------------
+; GETERROR: Return the error code in .A, and clear it.
+
+GETERROR:
+        lda     ERROR
+        stz     ERROR
+        rts
+
+; ------------------------------------------------------------------------
+; CONTROL: Platform-/driver-specific entry point.
+;
+; Must set an error code: YES
+
+CONTROL:
+        lda     #TGI_ERR_INV_FUNC
+        sta     ERROR
+        rts
+
+; ------------------------------------------------------------------------
+; CLEAR: Clear the screen.
+;
+; Must set an error code: NO
+
+CLEAR:
+    .scope inner
+
```

Comment:
Please use the same indention level for code in the entire file

Reply:
Got it!